From 0f4f6995d7e51f0f6cb17f1fcbe60efd8c684f36 Mon Sep 17 00:00:00 2001
From: "Pedro P. Candel" <pedro@joyent.com>
Date: Thu, 18 Aug 2016 19:01:18 +0200
Subject: [PATCH] TOOLS-1521 sdcadm should avoid updating duplicated agent
 instances on the same CN Reviewed by: Trent Mick <trent.mick@joyent.com>

---
 lib/procedures/update-agent-v1.js | 22 +++++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/lib/procedures/update-agent-v1.js b/lib/procedures/update-agent-v1.js
index 3190ebd..83aa868 100644
--- a/lib/procedures/update-agent-v1.js
+++ b/lib/procedures/update-agent-v1.js
@@ -5,7 +5,7 @@
  */
 
 /*
- * Copyright (c) 2015, Joyent, Inc.
+ * Copyright 2016, Joyent, Inc.
  */
 
 
@@ -156,6 +156,26 @@ UpdateAgentV1.prototype.execute = function uaExecute(opts, callback) {
                 });
             },
 
+            /*
+             * Guard against duplicate agent instance entries (see TOOLS-1521)
+             */
+            function preventDuplicateInstances(_, next) {
+                var instUUIDs = change.insts.map(function (ins) {
+                    return ins.instance;
+                });
+
+                var duplicates = instUUIDs.some(function (ins, pos) {
+                    return (instUUIDs.indexOf(ins) !== pos);
+                });
+
+                if (duplicates.length) {
+                    return next(new UpdateError(format(
+                        'there are duplicated instances for agent %s',
+                        change.insts[0].service)));
+                }
+                next();
+            },
+
             function checkCNsAvailability(_, next) {
                 sdcadm.cnapi.listServers({
                     uuids: change.insts.map(function (i) {
-- 
2.21.0

